src/opt.c

```diff
@@ -103,6 +104,71 @@ bool opt_get_verbose(opt_name_t name, const char *filter)
       return strstr(filter, value) != NULL;
 }
 
+int opt_parse_comma_separated(const char *opt, const char *optarg,
```

Comment:
There's really no need to make a separate generic function for this. It would be much simpler, easier to understand, and more correct if there was just a single function in `nvc.c` where we have `parse_cover_mask` now:

```c
void parse_cover_options(cover_mask_t *mask, int *array_limit);
```

That would completely avoid the need for the special "wildcard" handling as you could parse `ignore-arrays-from` specially.

Reply:
I would keep this one generic, since I think it can be usefull for other coverage unrelated options (thats why I changed it to be generic). E.g. from: https://github.com/nickg/nvc/issues/533

I can imagine something like:
`--dump-objs=signals,ports,memories,psl,constants,variables`

Reply:
I would just write a simple non-generic function like this:

```c
static void parse_cover_options(const char *str, cover_mask_t *mask,
                                int *array_limit)
{
   static const struct {
      const char *name;
      cover_mask_t mask;
   } options[] = {
      { "statement",             COVER_MASK_STMT                        },
      { "toggle",                COVER_MASK_TOGGLE                      },
      { "branch",                COVER_MASK_BRANCH                      },
      { "all",                   COVER_MASK_ALL                         },
      { "count-from-undefined",  COVER_MASK_TOGGLE_COUNT_FROM_UNDEFINED },
      { "count-from-to-z",       COVER_MASK_TOGGLE_COUNT_FROM_TO_Z      },
      { "ignore-mems",           COVER_MASK_TOGGLE_IGNORE_MEMS          },
   };

   for (const char *start = str; ; str++){
      if (*str == ',' || *str == '\0') {
         if (strncmp(start, "ignore-arrays-from-", 19) == 0)
            *array_limit = atoi(start + 19);
         else {
            int pos = 0;
            for (; pos < ARRAY_LEN(options); pos++) {
               if (!strncmp(options[pos].name, start, str - start))
                  break;
            }

            if (pos == ARRAY_LEN(options)) {
               diag_t *d = diag_new(DIAG_FATAL, NULL);
               diag_printf(d, "unknown coverage type '%.*s'",
                           (int)(str - start), start);
               diag_hint(d, NULL, "valid coverage types are: \n"
                         "  statement\n"
                         "  toggle\n"
                         "  branch");
               diag_hint(d, NULL, "selected coverage types shall be "
                         "comma separated e.g $bold$--cover=toggle,branch$$");
               diag_emit(d);
               fatal_exit(EXIT_FAILURE);
            }
            else
               *mask |= options[pos].mask;
         }

         if (*str == '\0')
            break;

         start = str + 1;
      }
   }
}
```

This is clear and easy to read, and also has a more precise error message. If we need comma separated argument parsing elsewhere I would rather just implement it separately. The problem with making a generic function for this is that it will become more and more complicated as we add new special cases to handle different types of options (like the wildcard parsing here). I would much rather have two simple functions that duplicate some logic than one complicated/obfuscated shared function.

Reply:
OK then, I will add the specific function, but TBH in this case I think having generic function
would be better than duplicated logic (as long as we don't bloat it with 10 other features).
I understand that sometimes bit of duplication is better than bit of dependency, but I am
afraid to end up with couple of different option parsers, one for each command, and have
different behavior in seemingly equal situations.

E.g. if I look at VCS parsing of command line arguments, it differs very much for VHDL (`vhdlan`) and
for Verilog (`vlogan`) compilation commads. So e.g. passing `std_logic_vector` to VHDL generic, and
passing `logic[X:Y]` as Verilog parameter requires completely different command line options, different
way how to escape the values, etc... The resulting user experience is __pain__.

It originates from the fact that those were two different compilers long time ago, but still, I would like to avoid this kind of situation.